### PR TITLE
projection net, encoding net, layers, and initializers

### PR DESCRIPTION
**Pull request by hnyu**
It has a dependency on encoding networks #366. For all networks, I always assume tensor outer_dims=0 to keep the APIs and code maintenance simpler. If there are use cases of multi outer_dims, the user can write a wrapper class based on specific problems. Also assume no nested input/output tensor specs for the same reason.

### CODE REVIEW

File: alf/layers.py [hunk 1]
```diff
@@ -11,104 +11,179 @@
...
+@gin.configurable
+class FC(nn.Module):
+    def __init__(self,
+                 input_size,
+                 output_size,
+                 activation=identity,
+                 use_bias=True,
+                 kernel_init_gain=1.0,
+                 bias_init_value=0.0):
+        """A fully connected layer that's also responsible for activaiton and
```
Comment (emailweixu) [inline alf/layers.py]: activaiton => activation

File: alf/layers.py [hunk 2]
```diff
@@ -11,104 +11,179 @@
...
+        self._activation = activation
+        self._conv_trans2d = nn.ConvTranspose2d(
+            in_channels,
+            out_channels,
+            kernel_size,
+            stride=strides,
+            padding=padding,
+            bias=use_bias)
+        variance_scaling_init(
+            self._conv_trans2d.weight.data,
```
Comment (emailweixu) [inline alf/layers.py]: is weight.data transposed or not? If it is transposed, then the fanin fanout need to be switched.

File: alf/networks/encoding_networks.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
+# Copyright (c) 2020 Horizon Robotics. All Rights Reserved.
```
Comment (emailweixu) [inline alf/networks/encoding_networks.py]: This is in the other PR #366. You either make 1 PR or have separate changes in 2 PRs.

File: alf/networks/projection_networks.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+import alf.layers as layers
+from alf.tensor_specs import TensorSpec, BoundedTensorSpec
+
+
+def DiagMultivariateNormal(loc, scale_diag):
+    return td.Independent(td.Normal(loc, scale_diag), 1)
+
+
+class CategoricalProjectionNetwork(nn.Module):
```
Comment (emailweixu) [inline alf/networks/projection_networks.py]: tf_agent (Categorical|Normal)ProjectionNetwork are subclass of DistributionNetwork, which has a property output_spec. I'm not sure we will need it or not. Just keep it in mind that we may have to implement it later.

File: alf/networks/encoding_networks.py (hunk 3 above)
Comment (hnyu) [reply alf/networks/encoding_networks.py]: Yeah this PR depends on #366. I can remove the files, but it will make the review more difficult (you have to look at code in another PR).

File: alf/layers.py (hunk 2 above)
Comment (hnyu) [reply alf/layers.py]: It's already transposed

Comment (hnyu) [reply alf/layers.py]: Yeah it's transposed.

File: alf/networks/encoding_networks.py [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+                layers.Conv2D(
+                    input_channels,
+                    filters,
+                    kernel_size,
+                    activation=activation,
+                    strides=strides,
+                    padding=padding))
+            input_channels = filters
+
+    def output_shape(self, input_size):
```
Comment (emailweixu) [inline alf/networks/encoding_networks.py]: Need to correctly handle when kernel_size, strides, padding are tuples

File: alf/networks/encoding_networks.py [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+            H = (H1-1) * strides + HF - 2P
+
+        where H = output size, H1 = input size, HF = size of kernel, P = padding
+
+        Returns:
+            a tuple representing the output shape (C,H,W)
+        """
+        _, height, width = self._start_decoding_shape
+        for paras in self._transconv_layer_params:
+            filters, kernel_size, strides = paras[:3]
```
Comment (emailweixu) [inline alf/networks/encoding_networks.py]: handle tuple

File: alf/networks/encoding_networks.py [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+                 fc_layer_params=None,
+                 activation=layers.identity,
+                 last_layer_size=None,
+                 last_activation=None):
+        """Create an EncodingNetwork
+
+        This EncodingNetwork allows the last layer to have different settings
+        from the other layers.
+
+        Args:
```
Comment (emailweixu) [inline alf/networks/encoding_networks.py]: TODO: Handle nested input_tensor_spec

File: alf/networks/initializers.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            so we need to compensate for that. Only used when mode=="fan_in".
+        nonlinearity (str): one of ['linear', 'conv1d', 'conv2d', 'conv3d',
+            'conv_transpose1d', 'conv_transpose2d', 'conv_transpose3d',
+            'sigmoid', 'tanh', 'relu', 'leaky_relu', 'identity']
+        nonlinearity_param (float): additional parameter of the nonlinearity;
+            currently only used by 'leaky_relu' as the negative slope (pytorch
+            default 0.01)
+        transposed (bool): a flag indicating if the weight tensor has been
+            tranposed (e.g., nn.ConvTranspose2d). In that case, `fan_in` and
+            `fan_out` should be swapped.
```
Comment (emailweixu) [inline alf/networks/initializers.py]: there are return at line 96 and 103. Need Returns comment section

File: alf/networks/projection_networks.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+            self._std_transform = std_transform
+
+        self._means_projection_layer = layers.FC(
+            input_size,
+            action_spec.shape[0],
+            activation=activation,
+            kernel_init_gain=projection_output_init_gain)
+
+        # We always assume that std depends on states
+        self._std_projection_layer = layers.FC(
```
Comment (emailweixu) [inline alf/networks/projection_networks.py]: For many models, we don't use state dependent std

File: alf/networks/projection_networks_test.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+        embedding = input_spec.ones(outer_dims=(1, ))
+
+        net = CategoricalProjectionNetwork(
+            input_size=input_spec.shape[0],
+            num_actions=5,
+            logits_init_output_factor=0)
+        dist = net(embedding)
+        self.assertTrue(torch.all(dist.probs == 0.2))
+
+    def test_zero_mean_projection_net(self):
```
Comment (emailweixu) [inline alf/networks/projection_networks_test.py]: what is the meaning of this function?

File: alf/networks/encoding_networks.py (hunk 7 above)
Comment (hnyu) [reply alf/networks/encoding_networks.py]: I don't think it's necessary to complicate the APIs in general. The users can write subclasses for specific problems. For now, unnested tensor specs is the common case.

File: alf/networks/encoding_networks.py [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+                layers.Conv2D(
+                    input_channels,
+                    filters,
+                    kernel_size,
+                    activation=activation,
+                    strides=strides,
+                    padding=padding))
+            input_channels = filters
+
+    def output_shape(self, input_size):
```
Comment (le-horizon) [inline alf/networks/encoding_networks.py]: I'm wondering, can we pass a 0 tensor of input_size to forward() and get the output_size from the output tensor?

That'll probably reduce code redundancy and won't add much overhead?

If we do this, then we can have a very simple output_shape function in the top level class, and all our custom nn.Module's can just inherit without explicitly implementing it?

Comment (hnyu) [reply alf/networks/encoding_networks.py]: Calculating output_shape is usually required for conv networks because for other layers you will know its value during construction. As for why I didn't use forward to get the output_shape is mainly for testing purpose in encoding_networks_test.py. The code amount won't be much less by doing forward.

Comment (le-horizon) [reply alf/networks/encoding_networks.py]: I see.  There are already two Modules that need output shape in this PR.  I worry less about code length, and more about code maintenance in the future.  If using forward can solve the problem once and for all, then why not?

Comment (hnyu) [reply alf/networks/encoding_networks.py]: Because for encoding_networks_test.py, I compare the forwarding-computed shape with the maunally calculated one. It's basically a sanity check.

Comment (le-horizon) [reply alf/networks/encoding_networks.py]: Then that means we can remove that test if we use forward to compute output shape in the output_shape() function?

Comment (hnyu) [reply alf/networks/encoding_networks.py]: Then how do you check if the networks are implemented correctly?

Comment (le-horizon) [reply alf/networks/encoding_networks.py]: Maybe move the manual computation logic into the test if you want.  Otherwise, just get a few example gold standard output shapes into the test?

Comment (le-horizon) [reply alf/networks/encoding_networks.py]: The only downside I can think of is whenever output_shape() is called, especially when input size is large, gpu memory could rocket.  But it's hopefully just a one time cost.

Comment (hnyu) [reply alf/networks/encoding_networks.py]: Also I personally don't feel comfortable that we are doing forward under the hood without telling the users. The log (error) message, if any, will be quite confusing, just like what we have now for prepare_specs.

Comment (le-horizon) [reply alf/networks/encoding_networks.py]: Ok, fair enough.  We can always reconsider if needed.
